### ISSUE TITLE
Make assignment expressions (`=`) return a value

### ISSUE THREAD

**Issue by altered-dev** [godotengine/godot-proposals#1065]
**Describe the project you are working on:** I am working on a game _about shooting bubbles_ and each bubble has a certain random chance of getting a bomb or a lock or something else.

**Describe the problem or limitation you are having in your project:** The code for setting the bubble type looks not clean:
```
var type_chance := rand_range(0, 10)
if type_chance <= bomb_chance: bubble.type = 1
elif type_chance <= bomb_chance + lock_chance: bubble.type = 2
elif type_chance <= bomb_chance + lock_chance + coin_chance: bubble.type = 3
elif type_chance <= bomb_chance + lock_chance + coin_chance + gem_chance: bubble.type = 4
elif type_chance <= bomb_chance + lock_chance + coin_chance + gem_chance + blind_chance: bubble.type = 5
elif type_chance <= bomb_chance + lock_chance + coin_chance + gem_chance + blind_chance + timer_chance: bubble.type = 6
else: bubble.type = 0
```

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:** This will shorten (but probably will not make it very understandable) some chunks of code, and this feature already exists in C# and probably a lot more languages (but is not in Python)

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:** In my instance, I could rewrite my code something like this:
```
var type_chance := rand_range(0, 10)
var type_comp := bomb_chance
if type_chance <= type_comp: bubble.type = 1
elif type_chance <= type_comp += lock_chance: bubble.type = 2
elif type_chance <= type_comp += coin_chance: bubble.type = 3
elif type_chance <= type_comp += gem_chance: bubble.type = 4
elif type_chance <= type_comp += blind_chance: bubble.type = 5
elif type_chance <= type_comp += timer_chance: bubble.type = 6
else: bubble.type = 0
```
It will also allow to make stuff like this:
```
some_var = other_var = third_var = forth_var
```
This will set all the variables to `forth_var`.
If I write this code right now, I get an error "Unexpected assign".

**If this enhancement will not be used often, can it be worked around with a few lines of script?:** When I posted this to Godot Q&A instead of this _on accident_, I was answered that I could make my code look like this:
```
var bomb_chance  := 0.2
var lock_chance  := bomb_chance  + 0.1
var coin_chance  := lock_chance  + 0.1
var gem_chance   := coin_chance  + 0.1
var blind_chance := gem_chance   + 0.1
var timer_chance := blind_chance + 0.1

var type_chance := rand_range(0, 1)

if type_chance <= bomb_chance:
    bubble.type = 1
elif type_chance <= lock_chance:
    bubble.type = 2
elif type_chance <= coin_chance:
    bubble.type = 3
elif type_chance <= gem_chance:
    bubble.type = 4
elif type_chance <= blind_chance:
    bubble.type = 5
elif type_chance <= timer_chance:
    bubble.type = 6
```

**Is there a reason why this should be core and not an add-on in the asset library?:** I pretty much don't know how to extend GDScript functionality, and I am not really sure if this possibility even exists.

**Comment by Calinou** [comment]
Allowing assignment in `if` statements is generally considered error-prone, as you could mistype `==` into `=`. This is the source of a lot of programming errors in languages like C and PHP.

**Comment by alt3r3dgd** [comment]
Maybe there could a warning exist saying "You wrote '=' here. Did you mean '=='?" or something like this.

**Comment by Calinou** [comment]
@alt3r3dgd I'm not sure if requiring people to write `#warning-ignore` comments is a good choice here (when you actually intend to assign with `=` in an `if` statement).

**Comment by alt3r3dgd** [comment]
Well, then there could be an option to enclose the assignment into parenthneses to remove the warning, this can also make the code more understandable 🤔

**Comment by KoBeWi** [comment]
> A well-written project shouldn't issue any GDScript warnings 🙂

Nah, some warnings are annoying and hard to avoid. I always just disable most of them. I imagine that would happen with this warning too. People who want to use condition assignment will just disable the warning project-wide, instead of doing it per use.

**Comment by dalexeev** [comment]
The example is unsuccessful, but the feature itself should be considered. Can anyone give examples when this feature really simplifies life?

**Comment by Koyper** [comment]
How about returning the assigned value, but only if the call is surrounded by parenthesis?

var, if and match, etc. not returning values is one of my only big complaints about the brilliant and lovely GDScript. I wrote 50k lines of complicated LISP where setq returns its value to an if statement so you can use the variable inside the scope of the if and also in the outer scope. Hundreds of cases of `cond` (`match`) returning its value to setq, `if` returning to setq...

That code would have been much longer and more complex if setq did not return its value, and I have hundreds of times longed for var to return it's value when defined.

Here's what is required now:

```
	var rgb64: int = _get_rgbint(color_info['color'])
	if not colors_index.has(rgb64):
		colors_info[rgb64] = color_info
	else:
		if not update_color:
			_remove_color_index_at_rgb(rgb64)
```

With var returning it's assignment that stays in the scope of the `if`:

```
	if not colors_index.has(var rgb64: int = _get_rgbint(color_info['color'])):
		colors_info[rgb64] = color_info
	else:
		if not update_color:
			_remove_color_index_at_rgb(rgb64)
```

Basically this pattern is annoying and cluttery, and the var predefinition can also drift lines above where readablility suffers.

```
var early_forced_definition = "something needed by multiple conditions of if or match"
if "How I wish I could be setting it right here" != early_forced_definition:
	early_forced_definition = "Grumble complaint to oneself and longsuffer"
else:
	early_forced_definition = "Thank you Godot devs, I bow to your brillance and generosity."

```

Becomes:
```
if  "How I wish I could be setting it right here" != (var early_forced_definition = "something needed by multiple conditions of if or match"):
	early_forced_definition = "Grumble complaint to oneself and longsuffer"
else:
	early_forced_definition = "Thank you Godot devs, I bow to your brillance and generosity."
```
This would be fantastic:

```
var if_only = (
	if a == b:
		d
	else:
		c
	)
	
```
And this too:
```
func _match_return(input: int) -> int:
	return (
		match input:
			a:
				1
			b:
				2
			c:
				3
	)
```
Instead of:
```
func _match_return(input: int) -> int:
	match input:
		a:
			return 1
		b:
			return 2
		c:
			return 3
```
It would be great to get rid of requiring redundant `return` calls.

**Comment by KoBeWi** [comment]
Related to #2727

**Comment by Zshandi** [comment]
> The example is unsuccessful, but the feature itself should be considered. Can anyone give examples when this feature really simplifies life?

Any time I need to clear multiple variables, such as by assigning them to 0:
```
variable1 = variable2 = variable3 = 0
```
instead of:
```
variable1 = 0
variable2 = 0
variable3 = 0
```

Or especially if they are strings or some other more complex data type and need to all have the same value set then it's even more helpful.

**Comment by AThousandShips** [comment]
The solution to that is to just use:
```gdscript
variable1 = 0
variable2 = variable1
variable3 = variable1
```

While I think it has its uses I feel that it's largely a code smell, if you need to assign multiple things the same value in *many places* that sounds like the code is unsound to begin with

To me chained assignments almost always hurt readability, statements in if statements are more valid IMO but I generally think they are also bad with a few exceptions

**Comment by HolonProduction** [comment]
This proposal has not managed to gather significant support over 6 years. I also heavily disagree with it from a language design standpoint. Taking both of this into account, I've decided to reject the proposal.
